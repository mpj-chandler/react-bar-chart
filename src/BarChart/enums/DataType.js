"use strict";
exports.__esModule = true;
var DataType;
(function (DataType) {
    DataType["Named"] = "NAMED";
    DataType["DateIndexed"] = "DATE_INDEXED";
    DataType["Numeric"] = "NUMERIC";
    DataType["NonNullNumeric"] = "NON_NULL_NUMERIC";
})(DataType || (DataType = {}));
;
exports["default"] = DataType;
