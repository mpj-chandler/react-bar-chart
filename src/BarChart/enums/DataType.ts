enum DataType {
    Named = 'NAMED',
    DateIndexed = 'DATE_INDEXED',
    Numeric = 'NUMERIC',
    NonNullNumeric = 'NON_NULL_NUMERIC',
};

export default DataType;
