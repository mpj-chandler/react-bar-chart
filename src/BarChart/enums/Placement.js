"use strict";
exports.__esModule = true;
var Placement;
(function (Placement) {
    Placement[Placement["Aligned"] = 0] = "Aligned";
    Placement[Placement["Bucket"] = 1] = "Bucket";
})(Placement || (Placement = {}));
exports["default"] = Placement;
