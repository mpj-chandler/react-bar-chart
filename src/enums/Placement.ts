enum Placement {
    Aligned,
    Bucket,
}

export default Placement;
