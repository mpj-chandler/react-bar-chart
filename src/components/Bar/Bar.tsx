import React from 'react';
import { useAnimation } from '../../utils/hooks/useAnimation/useAnimation';
import { Placement } from '../../enums/Placement';
import { NonNullNumericDataPoint, NamedDataPoint, DateIndexedDataPoint, DataPoint } from '../../__types__/seriesTypes';
import { AxisRange } from '../../__types__/axisTypes';
import { Padding } from '../../__types__/styling';
import { getBarXPosition } from './utils/getBarXPosition';
import DataType from '../../enums/DataType';
import AnimationEasingType from '../../enums/AnimationEasingFunction';

export interface BarProps {
    point: NonNullNumericDataPoint | NamedDataPoint | DateIndexedDataPoint;
    index: number;
    seriesIndex: number;
    dataType: DataType;
    yRange: AxisRange;
    xRange: AxisRange;
    width: number;
    stroke: string;
    padding: Padding;
    numBars: number;
    numSeries: number;
    placement: Placement;
    fillFormatter?: (seriesIndex: number, dataPoint: DataPoint, index: number) => string;
    animationEasingType?: AnimationEasingType;
}

function getBarEndingHeight(props: BarProps): number {
    const yMin = Math.min(0, Math.max(0, props.yRange.min));
    const totalVerticalPadding = props.padding.top + props.padding.bottom;

    if (props.yRange.max <= 0) {
        return (100 - totalVerticalPadding) * ((yMin - props.point.y1) / (props.yRange.min - props.yRange.max));
    }

    return (100 - totalVerticalPadding) * ((props.point.y1 - yMin) / (props.yRange.max - props.yRange.min));
}

function getZeroPoint(props: BarProps): number {
    return (props.yRange.max / (props.yRange.max - props.yRange.min));
}

function getY(height: number, yStart: number, value: number) {

    if (value >= 0) {
        return yStart - height;
    }

    return yStart;
}

const defaultProps: Partial<BarProps> = {
    animationEasingType: AnimationEasingType.None,
    fillFormatter: (seriesIndex: number, dataPoint: DataPoint, index: number) => '#FFF',
};

const Bar: React.FC<BarProps> = (props: BarProps) => {

    const adjProps = { ...defaultProps, ...props };

    const x = getBarXPosition(adjProps);
    const height = getBarEndingHeight(adjProps);
    const zeroPoint = getZeroPoint(adjProps);
    const animation = adjProps.animationEasingType === AnimationEasingType.None ? 1 : useAnimation(adjProps.animationEasingType, 600, 0);
    const yStart = (() => {
        if (adjProps.yRange.max <= 0) {
            return adjProps.padding.top;
        }

        return ((100 - (adjProps.padding.top + adjProps.padding.bottom)) * zeroPoint) + adjProps.padding.top;
    })();

    const fill = adjProps.fillFormatter(adjProps.seriesIndex, adjProps.point, adjProps.index);

    return (
        <rect
            x={`${x - adjProps.width / 2}%`}
            y={`${getY(animation * height, yStart, adjProps.point.y1)}%`}
            height={`${adjProps.point.y1 < 0 ? -1 * animation * height : animation * height}%`}
            width={`${adjProps.width / adjProps.numSeries}%`}
            fill={fill}
            stroke={adjProps.stroke}
        >
        </rect>
    );
}

export default Bar;
