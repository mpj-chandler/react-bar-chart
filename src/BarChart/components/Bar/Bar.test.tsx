import React from 'react';
import TestRenderer from 'react-test-renderer';
import Bar, { BarProps } from './Bar';
import Placement from '../../enums/Placement';
import DataType from '../../enums/DataType';

jest.useFakeTimers();

describe('Bar', () => {

    let container: HTMLDivElement | null;

    beforeEach(() => {
        container = document.createElement('div');
        document.body.appendChild(container);
    });

    afterEach(() => {
        if (container !== null) {
            document.body.removeChild(container);
            container = null;
        }
    });

    describe('When dealing with numeric data types', () => {
        describe('When all values on the yAxis are positive', () => {
            const props: BarProps = {
                point: { x: 10, y: 100 },
                index: 1,
                dataType: DataType.NonNullNumeric,
                xRange: { min: 0, max: 100 },
                yRange: { min: 0, max: 200 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                placement: Placement.Aligned,
                numBars: 3,
                seriesIndex: 0,
                numSeries: 1,
            };

            it('it renders consistently with defaults', () => {
                checkSnapshot(props);
            });

            it('it renders consistently with Placement set to Aligned', () => {
                checkSnapshot({ ...props, placement: Placement.Aligned });
            });

            it('it renders consistently with Placement set to Bucket', () => {
                checkSnapshot({ ...props, placement: Placement.Bucket });
            });
        });

        describe('When all values on the yAxis are negative', () => {
            const props: BarProps = {
                point: { x: 10, y: -50 },
                index: 1,
                dataType: DataType.NonNullNumeric,
                xRange: { min: 0, max: 100 },
                yRange: { min: -100, max: -100 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                placement: Placement.Aligned,
                numBars: 3,
                seriesIndex: 0,
                numSeries: 1,
            };

            it('it renders consistently with defaults', () => {
                checkSnapshot(props);
            });

            it('it renders consistently with Placement set to Aligned', () => {
                checkSnapshot({ ...props, placement: Placement.Aligned });
            });

            it('it renders consistently with Placement set to Bucket', () => {
                checkSnapshot({ ...props, placement: Placement.Bucket });
            });
        });


        describe('When the yAxis spans positive and negative values', () => {
            const props: BarProps = {
                index: 1,
                dataType: DataType.NonNullNumeric,
                point: { x: 10, y: -50 },
                xRange: { min: 0, max: 100 },
                yRange: { min: -100, max: 100 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                placement: Placement.Aligned,
                numBars: 3,
                seriesIndex: 0,
                numSeries: 1,
            };

            it('it renders consistently with defaults', () => {
                checkSnapshot(props);
            });

            it('it renders consistently with Placement set to Aligned', () => {
                checkSnapshot({ ...props, placement: Placement.Aligned });
            });

            it('it renders consistently with Placement set to Bucket', () => {
                checkSnapshot({ ...props, placement: Placement.Bucket });
            });
        });
    });

    describe('When dealing with named data types', () => {
        describe('When all values on the yAxis are positive', () => {
            const props: BarProps = {
                point: { x: 'Anything', y: 100 },
                index: 1,
                dataType: DataType.Named,
                xRange: { min: 0, max: 100 },
                yRange: { min: 0, max: 200 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                placement: Placement.Aligned,
                numBars: 3,
                seriesIndex: 0,
                numSeries: 1,
            };

            it('it renders consistently with defaults', () => {
                checkSnapshot(props);
            });

            it('it renders consistently with Placement set to Aligned', () => {
                checkSnapshot({ ...props, placement: Placement.Aligned });
            });

            it('it renders consistently with Placement set to Bucket', () => {
                checkSnapshot({ ...props, placement: Placement.Bucket });
            });
        });

        describe('When all values on the yAxis are negative', () => {
            const props: BarProps = {
                point: { x: 'Anything', y: -50 },
                index: 1,
                dataType: DataType.Named,
                xRange: { min: 0, max: 100 },
                yRange: { min: -100, max: -100 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                placement: Placement.Aligned,
                numBars: 3,
                seriesIndex: 0,
                numSeries: 1,
            };

            it('it renders consistently with defaults', () => {
                checkSnapshot(props);
            });

            it('it renders consistently with Placement set to Aligned', () => {
                checkSnapshot({ ...props, placement: Placement.Aligned });
            });

            it('it renders consistently with Placement set to Bucket', () => {
                checkSnapshot({ ...props, placement: Placement.Bucket });
            });
        });


        describe('When the yAxis spans positive and negative values', () => {
            const props: BarProps = {
                index: 1,
                dataType: DataType.Named,
                point: { x: 'Anything', y: -50 },
                xRange: { min: 0, max: 100 },
                yRange: { min: -100, max: 100 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                placement: Placement.Aligned,
                numBars: 3,
                seriesIndex: 0,
                numSeries: 1,
            };

            it('it renders consistently with defaults', () => {
                checkSnapshot(props);
            });

            it('it renders consistently with Placement set to Aligned', () => {
                checkSnapshot({ ...props, placement: Placement.Aligned });
            });

            it('it renders consistently with Placement set to Bucket', () => {
                checkSnapshot({ ...props, placement: Placement.Bucket });
            });
        });
    });

    describe('When attempting a render with an unknown data type', () => {
        it('it throws an error', () => {
            const props: BarProps = {
                index: 1,
                dataType: DataType.DateIndexed,
                point: { x: 'Anything', y: -50 },
                xRange: { min: 0, max: 100 },
                yRange: { min: -100, max: 100 },
                width: 100,
                fill: 'anything',
                stroke: 'anything',
                padding: {
                    top: 10,
                    left: 10,
                    right: 10,
                    bottom: 10,
                },
                placement: Placement.Aligned,
                numBars: 3,
                seriesIndex: 0,
                numSeries: 1,
            };

            expect(() => TestRenderer.create(<Bar {...props} placement={Placement.Aligned}/>)).toThrowError('Error in rendering bar! Unrecognized data type!');
        });
    });
});

function checkSnapshot(props: BarProps) {
    const component = TestRenderer.create(<Bar {...props}/>);
    expect(component.toJSON()).toMatchSnapshot();
}

