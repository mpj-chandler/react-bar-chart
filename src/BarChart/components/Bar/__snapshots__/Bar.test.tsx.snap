// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar When dealing with named data types When all values on the yAxis are negative it renders consistently with Placement set to Aligned 1`] = `
<rect
  fill="anything"
  height="NaN%"
  stroke="anything"
  width="100%"
  x="-39.2%"
  y="10%"
/>
`;

exports[`Bar When dealing with named data types When all values on the yAxis are negative it renders consistently with Placement set to Bucket 1`] = `
<rect
  fill="anything"
  height="NaN%"
  stroke="anything"
  width="100%"
  x="-25.866666666666667%"
  y="10%"
/>
`;

exports[`Bar When dealing with named data types When all values on the yAxis are negative it renders consistently with defaults 1`] = `
<rect
  fill="anything"
  height="NaN%"
  stroke="anything"
  width="100%"
  x="-39.2%"
  y="10%"
/>
`;

exports[`Bar When dealing with named data types When all values on the yAxis are positive it renders consistently with Placement set to Aligned 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-39.2%"
  y="90%"
/>
`;

exports[`Bar When dealing with named data types When all values on the yAxis are positive it renders consistently with Placement set to Bucket 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-25.866666666666667%"
  y="90%"
/>
`;

exports[`Bar When dealing with named data types When all values on the yAxis are positive it renders consistently with defaults 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-39.2%"
  y="90%"
/>
`;

exports[`Bar When dealing with named data types When the yAxis spans positive and negative values it renders consistently with Placement set to Aligned 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-39.2%"
  y="50%"
/>
`;

exports[`Bar When dealing with named data types When the yAxis spans positive and negative values it renders consistently with Placement set to Bucket 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-25.866666666666667%"
  y="50%"
/>
`;

exports[`Bar When dealing with named data types When the yAxis spans positive and negative values it renders consistently with defaults 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-39.2%"
  y="50%"
/>
`;

exports[`Bar When dealing with numeric data types When all values on the yAxis are negative it renders consistently with Placement set to Aligned 1`] = `
<rect
  fill="anything"
  height="NaN%"
  stroke="anything"
  width="100%"
  x="-32%"
  y="10%"
/>
`;

exports[`Bar When dealing with numeric data types When all values on the yAxis are negative it renders consistently with Placement set to Bucket 1`] = `
<rect
  fill="anything"
  height="NaN%"
  stroke="anything"
  width="100%"
  x="-18.666666666666668%"
  y="10%"
/>
`;

exports[`Bar When dealing with numeric data types When all values on the yAxis are negative it renders consistently with defaults 1`] = `
<rect
  fill="anything"
  height="NaN%"
  stroke="anything"
  width="100%"
  x="-32%"
  y="10%"
/>
`;

exports[`Bar When dealing with numeric data types When all values on the yAxis are positive it renders consistently with Placement set to Aligned 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-32%"
  y="90%"
/>
`;

exports[`Bar When dealing with numeric data types When all values on the yAxis are positive it renders consistently with Placement set to Bucket 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-18.666666666666668%"
  y="90%"
/>
`;

exports[`Bar When dealing with numeric data types When all values on the yAxis are positive it renders consistently with defaults 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-32%"
  y="90%"
/>
`;

exports[`Bar When dealing with numeric data types When the yAxis spans positive and negative values it renders consistently with Placement set to Aligned 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-32%"
  y="50%"
/>
`;

exports[`Bar When dealing with numeric data types When the yAxis spans positive and negative values it renders consistently with Placement set to Bucket 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-18.666666666666668%"
  y="50%"
/>
`;

exports[`Bar When dealing with numeric data types When the yAxis spans positive and negative values it renders consistently with defaults 1`] = `
<rect
  fill="anything"
  height="0%"
  stroke="anything"
  width="100%"
  x="-32%"
  y="50%"
/>
`;
